File: servers/x/src/throttle.ts
/**
 * Pacing and backoff for every tool call that drives the browser.
 *
 * The write behaviours already wait between individual clicks; nothing spaced
 * the tool calls themselves, and a stream of scrapes with no gap is the pattern
 * that gets an account flagged. `pace` puts a jittered floor under the interval
 * between calls. `penalize` is what the 429 listener was missing: X was already
 * being detected, and the server did nothing with it beyond reporting.
 */

const PACE_MIN_MS = 5_000;
const PACE_JITTER_MS = 5_000;
// The first step has to be short enough to sit out inside one `wait` call.
const BACKOFF_BASE_MS = 30_000;
const BACKOFF_MAX_MS = 15 * 60_000;
/** A single wait call sleeps at most this long; longer cooldowns take several. */
const WAIT_CAP_MS = 45_000;
/** Blocked this many times in one cooldown and we stop waiting — hanging is not a plan. */
const MAX_BLOCKED = 3;

export class Throttle {
  private last = 0;
  private until = 0;
  private strikes = 0;
  private blocked = 0;

  /** Blocks until enough time has passed since the previous call. */
  async pace(): Promise<void> {
    const gap = PACE_MIN_MS + Math.floor(Math.random() * PACE_JITTER_MS);
    const wait = this.last === 0 ? 0 : this.last + gap - Date.now();
    this.last = Date.now() + Math.max(0, wait);
    if (wait > 0) await new Promise((resolve) => setTimeout(resolve, wait));
  }

  /**
   * The remaining cooldown, or null. This reports; it does not sit and wait —
   * waiting is the `wait` tool's job, so the caller gets to decide between
   * waiting it out and going to do something else.
   */
  check(): string | null {
    const remaining = this.until - Date.now();
    if (remaining <= 0) {
      this.blocked = 0; // this cooldown is over
      return null;
    }

    this.blocked += 1;
    const secs = Math.ceil(remaining / 1000);
    if (this.blocked >= MAX_BLOCKED) {
      return (
        `Waiting out X's cooldown did not work: blocked ${this.blocked} times, ${secs}s ` +
        `still left. Stop waiting and report this read as failed — whether the caller ` +
        `still wants it is their call.`
      );
    }
    return (
      `Rate limited by X — ${secs}s of cooldown left (${this.strikes} in a row, ` +
      `blocked ${this.blocked}/${MAX_BLOCKED}). X is not gone, just unusable right now: ` +
      `call wait to sit it out, then retry.`
    );
  }

  /**
   * Sleeps off up to WAIT_CAP_MS of the cooldown. Chunked because one tool call
   * cannot block indefinitely; if time remains, the caller just calls again.
   */
  async cool(): Promise<{ slept: number; remaining: number }> {
    const remaining = this.until - Date.now();
    // Once we have given up, wait must stop sleeping, or it becomes the place
    // the whole thing hangs.
    if (remaining <= 0 || this.blocked >= MAX_BLOCKED) {
      return { slept: 0, remaining: Math.max(0, remaining) };
    }

    const slept = Math.min(remaining, WAIT_CAP_MS);
    await new Promise((resolve) => setTimeout(resolve, slept));
    return { slept, remaining: remaining - slept };
  }

  penalize(): void {
    this.strikes += 1;
    const wait = Math.min(BACKOFF_BASE_MS * 2 ** (this.strikes - 1), BACKOFF_MAX_MS);
    this.until = Date.now() + wait;
  }

  succeed(): void {
    this.strikes = 0;
    this.blocked = 0;
    this.until = 0;
  }
}

File: servers/linkedin/linkedin_mcp_server/throttle.py
"""Pacing and backoff for every tool call that drives the browser.

`SequentialToolExecutionMiddleware` already guarantees that only one tool call
touches the browser at a time. What it does not do is leave a gap between them,
and a burst of scrapes with no gap is the pattern that gets an account
restricted. That matters more here than it looks: reading one feed is a couple
of page loads, but a run that searches ten queries three pages deep and then
opens twenty profiles is closer to seventy-five — a different order of
magnitude against the same account.

Rate limiting was detected and then thrown away. `detect_rate_limit` raises,
`_RATE_LIMITED_MSG` comes back for the soft case, the extractor retries once
after five seconds — and the next call goes straight back in as if nothing had
happened. `penalize` is what makes being blocked change the server's behaviour
rather than only its error message.

There is no `wait` tool to sit out a cooldown with, and sleeping through one
inside a call would hit the 180s tool timeout anyway. So a cooling server
refuses instead: a caller that stops for the day is doing the right thing, and
retrying into a restriction is how a warning becomes a ban.
"""

from __future__ import annotations

import asyncio
import logging
import random
import time

import mcp.types as mt

from fastmcp.exceptions import ToolError
from fastmcp.server.middleware import CallNext, Middleware, MiddlewareContext
from fastmcp.tools import ToolResult

from linkedin_mcp_server.core.exceptions import RateLimitError

logger = logging.getLogger(__name__)

# LinkedIn is stricter about scraping than the other platforms here, and this
# floor is the main thing standing between a long run and a restricted account.
# Sixty tool calls at ~9s average is about nine minutes of pure waiting, which
# is a price worth paying for a job nobody is waiting on.
PACE_MIN_SECONDS = 6.0
PACE_JITTER_SECONDS = 6.0

# Deliberately harsher than a rate limit deserves, because the failure it is
# guarding against is not a slow run — it is losing the account.
BACKOFF_BASE_SECONDS = 60.0
BACKOFF_MAX_SECONDS = 30 * 60.0


class Throttle:
    """Paces browser-driving tool calls and backs off when LinkedIn pushes back."""

    def __init__(self) -> None:
        self._last = 0.0
        self._until = 0.0
        self._strikes = 0

    async def pace(self) -> None:
        """Block until enough time has passed since the previous call."""
        gap = PACE_MIN_SECONDS + random.random() * PACE_JITTER_SECONDS
        now = time.monotonic()
        wait = 0.0 if self._last == 0.0 else self._last + gap - now
        # Claim the slot before sleeping so concurrent waiters cannot all
        # measure the gap from the same starting point.
        self._last = now + max(0.0, wait)
        if wait > 0:
            await asyncio.sleep(wait)

    def cooling(self) -> str | None:
        """The reason to refuse this call, or None when it may proceed."""
        remaining = self._until - time.monotonic()
        if remaining <= 0:
            return None
        return (
            f"LinkedIn rate-limited this session — {int(remaining)}s of cooldown left "
            f"({self._strikes} in a row). This server will not retry into a "
            f"restriction: stop reading LinkedIn for now and report the run as "
            f"incomplete rather than working around this."
        )

    def penalize(self) -> None:
        self._strikes += 1
        wait = min(BACKOFF_BASE_SECONDS * 2 ** (self._strikes - 1), BACKOFF_MAX_SECONDS)
        self._until = time.monotonic() + wait
        logger.warning(
            "LinkedIn rate limit #%d — refusing tool calls for %.0fs",
            self._strikes,
            wait,
        )

    def succeed(self) -> None:
        self._strikes = 0
        self._until = 0.0


class ThrottleMiddleware(Middleware):
    """Applies the throttle to every tool call that drives the browser.

    Registered after ``SequentialToolExecutionMiddleware`` so it runs inside
    that lock: pacing outside the lock would let calls queue up and then fire
    back to back the moment it released, which is the burst it exists to stop.
    """

    # Nothing here touches LinkedIn, so nothing here should wait or be refused.
    _EXEMPT = frozenset({"close_session"})

    def __init__(self, throttle: Throttle | None = None) -> None:
        self._throttle = throttle or Throttle()

    async def on_call_tool(
        self,
        context: MiddlewareContext[mt.CallToolRequestParams],
        call_next: CallNext[mt.CallToolRequestParams, ToolResult],
    ) -> ToolResult:
        tool_name = context.message.name
        if tool_name in self._EXEMPT:
            return await call_next(context)

        cooling = self._throttle.cooling()
        if cooling:
            raise ToolError(cooling)

        await self._throttle.pace()

        try:
            result = await call_next(context)
        except RateLimitError:
            # Detected here since the beginning and acted on by nothing.
            self._throttle.penalize()
            raise

        # The soft case never raises: the extractor swallows it, retries once,
        # and hands back a sentinel as if it were content. Left unread, the run
        # keeps hammering a session LinkedIn has already started refusing.
        if "[Rate limited]" in str(result.structured_content):
            self._throttle.penalize()
            logger.warning("Soft rate limit in '%s' result", tool_name)
            return result

        self._throttle.succeed()
        return result
